cmake/Boost.cmake

```diff
@@ -0,0 +1,93 @@
+macro(p4c_obtain_boost)
+  option(
+    P4C_USE_PREINSTALLED_BOOST
+    "Look for a preinstalled version of Boost in the system instead of installing the library using FetchContent."
+    OFF
+  )
+
+  # If P4C_USE_PREINSTALLED_BOOST is ON just try to find a preinstalled version of Boost.
+  if(P4C_USE_PREINSTALLED_BOOST)
+    if(NOT BUILD_SHARED_LIBS)
+      # Link Boost statically.
+      # See https://cmake.org/cmake/help/latest/module/FindBoost.html for details.
+      set(Boost_USE_STATIC_LIBS ON)
+      set(Boost_USE_STATIC_RUNTIME OFF)
+    endif()
+
+    # The boost graph headers are optional and only required by the graphs back end.
+    find_package(Boost QUIET COMPONENTS graph)
+    if(Boost_FOUND)
+      set(HAVE_LIBBOOST_GRAPH 1)
+    else()
+      message(WARNING "Boost graph headers not found, will not build 'graphs' backend")
+    endif()
+    find_package(Boost REQUIRED COMPONENTS iostreams)
+    set(P4C_BOOST_LIBRARIES Boost::iostreams)
+  else()
+    # This option can be used to include custom boost modules.
+    set(P4C_TARGET_BOOST_LIBRARIES "" CACHE STRING "Build and make these boost modules available. Modules are separated with semicolons.")
+    set(P4C_BOOST_VERSION "1.86.0")
+    message(STATUS "Fetching Boost version ${P4C_BOOST_VERSION} for P4C...")
+    # Print out download state while setting up Boost.
+    set(FETCHCONTENT_QUIET_PREV ${FETCHCONTENT_QUIET})
+    set(FETCHCONTENT_QUIET OFF)
+    # Unity builds do not work for Boost...
+    set(CMAKE_UNITY_BUILD_PREV ${CMAKE_UNITY_BUILD})
+    set(CMAKE_UNITY_BUILD OFF)
+
+    # Add boost modules.
+    # format, multiprecision, and iostreams are needed by P4C core.
+    set(BOOST_INCLUDE_LIBRARIES format multiprecision iostreams)
```

Comment:
This is a variable that is consumed by the Boost CMakefiles to determine which libraries to enable. I think it could be specified by the caller but those three libraries are required.

Reply:
Consider the following case (take `graph` backend as a model example): let us assume there is a backend that would require more boost modules. How these could be accomplished? It seems there should be some generic way to specify boost libraries that are to be used beyond the ones required by frontend currently.

Reply:
Yes, `P4C_TARGET_BOOST_LIBRARIES` is intended for that. You either specify that at the command line or before invoking the Boost CMake or p4c_obtain_boost. Unfortunately,  a back end can not retroactively add dependencies. Or at least I have not tried doing that. 

I could make this a function parameter but that still has the same limitations from what I can see.

Reply:
How is `P4C_TARGET_BOOST_LIBRARIES` related to `ADDITIONAL_P4C_BOOST_LIBRARIES`? Also what exactly does "on commandline" mean? Ideally I would want to be able to set it from downstreams top-level cmake *before* including `p4c`. Or at least from cmake presets for the downstream.

Reply:
In that case it should work well. I was confused by "on commandline".

Reply:
Yeah both options are possible. You could also just set the required modules using `cmake -DADDITIONAL_P4C_BOOST_LIBRARIES=assign;filesystem;graph;..."